File: leet-code/435.无重叠区间.js
/*
 * @lc app=leetcode.cn id=435 lang=javascript
 *
 * [435] 无重叠区间
 */

// @lc code=start
/**
 * @param {number[][]} intervals
 * @return {number}
 */
var eraseOverlapIntervals = function (intervals) {
  intervals.sort((a, b) => {
    if (a[0] === b[0]) {
      return a[1] - b[1]
    } else {
      return a[0] - b[0]
    }
  })
  let i = 0
  let res = 0
  while (i < intervals.length) {
    const point1 = intervals[i]
    let right = point1[1]
    let j = i + 1
    while (j < intervals.length) {
      const point2 = intervals[j]
      if (point2[0] >= right) break
      j++
      // 更新右边界
      right = Math.min(right, point2[1])
    }
    res++
    i = j
  }
  return intervals.length - res
};
// @lc code=end

File: leet-code-rs/src/solutions/435_无重叠区间.rs
/*
 * @lc app=leetcode.cn id=435 lang=rust
 *
 * [435] 无重叠区间
 */

// @lc code=start

#[allow(unused)]
impl Solution {
    pub fn erase_overlap_intervals(mut intervals: Vec<Vec<i32>>) -> i32 {
        intervals.sort_by(|a, b| a[0].cmp(&b[0]));

        let mut res = 0;
        for i in 1..intervals.len() {
            if intervals[i][0] < intervals[i - 1][1] {
                // 有交集
                res += 1;
                // 更新右区间
                intervals[i][1] = intervals[i][1].min(intervals[i - 1][1]);
            }
        }

        return res;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_erase_overlap_intervals() {
        //assert_eq!(Solution::erase_overlap_intervals(vec![]),[]);
        assert!(true)
    }
}
